amads/time/meter/__init__.py

```diff
@@ -0,0 +1,619 @@
+    # 2. "hidden" layer/s
+    if len(numerators) > 1:  # whole cycle covered.
+        if isinstance(numerators[0], list):
+            for level in numerators:
+                starts = starts_from_beat_pattern(level, denominator)
+                start_hierarchy.append(starts)
+        else:
+            starts = starts_from_beat_pattern(numerators, denominator)
+            start_hierarchy.append(starts)
+
+    # 3. Finally, everything at the denominator level and shorter:
+    this_denominator_power = supported_denominators.index(denominator)
+    max_denominator_power = supported_denominators.index(minimum_pulse)
+
+    for this_denominator in supported_denominators[
+        this_denominator_power : max_denominator_power + 1
+    ]:
+        this_ql = 4 / this_denominator
+        starts = starts_from_lengths(measure_length, this_ql)
+        start_hierarchy.append(starts)
+
+    return start_hierarchy
+
+
+def starts_from_time_signature_and_levels(
+    ts_str: str,
+    levels: Union[list, None] = None,
+):
+    """
+    Gets starts from a time signature and a list of levels.
+    Records the starts associated with each level as a list,
+    and return a list of those lists.
+
+    Levels are defined by the hierarchies recorded in the time signature.
+    The 0th level for the full cycle (start 0.0 and full measure length) is always included.
+    "1" stands for the 1st level below that of the full metrical cycle
+    (e.g., a half cycle of 4/4 and similar time signtures like 2/2).
+    "2" is the next level down, (quarter cycle) and so on.
+
+    The function arranges a list of levels in increasing order:
+    level 0 is always included, and
+    the maximum permitted level (depth) is 6.
+
+    Level choices do not need to be successive:
+    e.g., in 4/4 a user could choose [1, 3],
+    with 1 standing for the half note level,
+    3 for the eight note level,
+    and skipping in intervening level 2 (quarter note).
+
+    Parameters
+    ----------
+    ts_str
+        Any valid string repersenting a time signature. See examples below.
+    levels
+        Optionally specify which levels implied by the time signature to use.
+
+    Returns
+    -------
+    list
+        Returns a list of lists with start positions by level.
+
+    Examples
+    --------
+
+    >>> starts_from_time_signature_and_levels("6/8", [1, 2])
+    [[0.0, 3.0], [0.0, 1.5, 3.0], [0.0, 0.5, 1.0, 1.5, 2.0, 2.5, 3.0]]
+
+    >>> level_not_specified = starts_from_time_signature_and_levels("6/8")
+    >>> len(level_not_specified)
+    6
+
+    >>> level_not_specified[3]
+    [0.0, 0.25, 0.5, 0.75, 1.0, 1.25, 1.5, 1.75, 2.0, 2.25, 2.5, 2.75, 3.0]
+
+    """
+    if levels is None:
+        levels = range(6)
+
+    if max(levels) > 6:
+        raise ValueError("6 is the maximum level depth supported.")
+    levels = sorted(levels)  # smallest first
+    if 0 not in levels:  # likely, and not a problem
+        levels = [0] + levels  # Always have level 0 as first entry
```

Comment:
If you refactor to implement `MetricalHierarchy.from_time_signature` as suggested above, then I think you could implement this quite succinctly as:

```py
    @classmethod
    def from_time_signature(cls, time_signature: str, levels: Optional[List[int]]):
        numerator, denominator = time_signature.split("/")
        ...
        hierarchy = ...
        levels = set([0] + levels)
        hierarchy = [pulses for level, pulses in hierarchy if depth in levels]
        return cls(hierarchy)
        ....
```

Reply:
Both the refactor and the use of `depth` are open questions moving to the issue.